src/aave-v3/SupplyVault.sol

```diff
@@ -217,36 +196,61 @@ contract SupplyVault is ISupplyVault, SupplyVaultBase {
             (rewardTokens, claimedAmounts) = morpho.claimRewards(poolTokens, false);
         }
 
-        uint256 supply = totalSupply();
         for (uint256 i; i < rewardTokens.length; ++i) {
             address rewardToken = rewardTokens[i];
             uint256 claimedAmount = claimedAmounts[i];
+            UserRewardsData storage userRewardsData = userRewards[rewardToken][_user];
             uint128 rewardsIndexMem;
 
-            if (supply > 0 && claimedAmount > 0) {
+            if (totalSupply() > 0 && claimedAmount > 0) {
                 rewardsIndexMem =
                     rewardsIndex[rewardToken] +
-                    claimedAmount.mulDivDown(SCALE, supply).safeCastTo128();
+                    _getUnaccruedRewardIndex(claimedAmount);
                 rewardsIndex[rewardToken] = rewardsIndexMem;
             } else rewardsIndexMem = rewardsIndex[rewardToken];
 
-            UserRewardsData storage userRewardsData = userRewards[rewardToken][_user];
-            uint256 rewardsIndexDiff;
-
-            // Safe because we always have `rewardsIndex` >= `userRewardsData.index`.
-            unchecked {
-                rewardsIndexDiff = rewardsIndexMem - userRewardsData.index;
-            }
-
-            if (rewardsIndexDiff > 0) {
-                uint128 accruedRewards = balanceOf(_user)
-                .mulDivDown(rewardsIndexDiff, SCALE)
-                .safeCastTo128();
-                userRewardsData.unclaimed += accruedRewards;
+            if (rewardsIndexMem > userRewardsData.index) {
+                uint128 accruedReward = _getUnaccruedRewardFromUnaccruedRewardIndex(
+                    _user,
+                    rewardsIndexMem - userRewardsData.index
+                );
+                userRewardsData.unclaimed += accruedReward;
                 userRewardsData.index = rewardsIndexMem;
 
-                emit Accrued(rewardToken, _user, rewardsIndexMem, accruedRewards);
+                emit Accrued(rewardToken, _user, rewardsIndexMem, accruedReward);
             }
         }
     }
+
+    function _getUnaccruedReward(
+        address _user,
+        address _rewardToken,
+        uint256 _claimableReward
+    ) internal view returns (uint128 unaccruedReward) {
+        unaccruedReward = _getUnaccruedRewardFromUnaccruedRewardIndex(
+            _user,
+            _getUnaccruedRewardIndex(_claimableReward) + // The unaccrued reward index
+                rewardsIndex[_rewardToken] -
+                userRewards[_rewardToken][_user].index // The difference between the current reward index and the user's index
+        );
+    }
+
+    function _getUnaccruedRewardFromUnaccruedRewardIndex(
+        address _user,
+        uint128 _unaccruedRewardIndex
+    ) internal view returns (uint128 unaccruedReward) {
+        unaccruedReward = balanceOf(_user).mulDivDown(_unaccruedRewardIndex, SCALE).safeCastTo128();
+    }
+
+    function _getUnaccruedRewardIndex(uint256 _claimableReward)
```

Comment:
I think it's appropriate because what is returned from this function is not the updated reward index, but instead the amount of reward index that the current reward index should be increased by.

Reply:
Oh yes sorry, this is the unaccrued that is a bit weird since I'm more considering "accrual" on a quantity of tokens and not an index